File: companion/bolumler/b09/ornek-01/cozum.R
hazirla <- function(veri) {
  if (!identical(names(veri), c("fark", "standart_hata", "serbestlik", "null_degeri", "alfa")) || nrow(veri) != 1) {
    stop("Tek satirda fark, standart_hata, serbestlik, null_degeri, alfa ozeti gerekli.")
  }
  if (!all(vapply(veri, is.numeric, logical(1))) || anyNA(veri) ||
      any(!is.finite(as.matrix(veri)))) stop("Sonlu ve eksiksiz sayisal ozet gerekli.")
  if (veri$standart_hata <= 0 || veri$serbestlik <= 0) stop("SE ve serbestlik pozitif olmali.")
  if (veri$alfa <= 0 || veri$alfa >= 1) stop("Alfa 0 ile 1 arasinda olmali.")
  veri
}

test_hesapla <- function(veri) {
  temiz <- hazirla(veri)
  t_degeri <- (temiz$fark - temiz$null_degeri) / temiz$standart_hata
  p_cift <- 2 * pt(abs(t_degeri), df = temiz$serbestlik, lower.tail = FALSE)
  p_ust <- pt(t_degeri, df = temiz$serbestlik, lower.tail = FALSE)
  p_alt <- pt(t_degeri, df = temiz$serbestlik)
  kritik <- qt(1 - temiz$alfa / 2, df = temiz$serbestlik)
  hata_payi <- kritik * temiz$standart_hata
  alt_sinir <- temiz$fark - hata_payi
  ust_sinir <- temiz$fark + hata_payi
  sonuc <- c(t = t_degeri, p_cift = p_cift, p_ust = p_ust, p_alt = p_alt,
              reddet_cift = as.integer(p_cift < temiz$alfa), reddet_alfa001 = as.integer(p_cift < 0.01),
              reddet_ust = as.integer(p_ust < temiz$alfa), reddet_alt = as.integer(p_alt < temiz$alfa),
              guven_duzeyi = 1 - temiz$alfa, kritik_t = kritik, hata_payi = hata_payi,
              alt_sinir = alt_sinir, ust_sinir = ust_sinir, genislik = 2 * hata_payi,
              null_aralikta = as.integer(alt_sinir <= temiz$null_degeri && temiz$null_degeri <= ust_sinir))
  if (any(!is.finite(sonuc))) stop("Test veya aralik sayisal olarak hesaplanamadi.")
  sonuc
}

hesapla <- function(veri) {
  temiz <- hazirla(veri)
  sonuc <- test_hesapla(temiz)
  data.frame(degisken = c(rep("girdi", 5), rep("test", 8), rep("aralik", 7)),
             olcu = c(names(temiz), names(sonuc)),
             deger = c(as.numeric(unlist(temiz)), as.numeric(sonuc)), stringsAsFactors = FALSE)
}

kontrol_et <- function(sonuc, beklenen) {
  if (!identical(names(beklenen), c("degisken", "olcu", "deger")) ||
      !identical(sonuc$degisken, beklenen$degisken) ||
      !identical(sonuc$olcu, beklenen$olcu)) stop("Kontrol etiketleri uyusmuyor.")
  if (!is.numeric(beklenen$deger) || anyNA(beklenen$deger) ||
      any(!is.finite(beklenen$deger)) || any(!is.finite(sonuc$deger))) stop("Sonlu kontrol gerekli.")
  tolerans <- pmax(1e-9, 1e-9 * pmax(abs(sonuc$deger), abs(beklenen$deger)))
  if (any(abs(sonuc$deger - beklenen$deger) > tolerans)) stop("Kontrol uyusmazligi.")
  cat("DOGRULANDI:", nrow(sonuc), "kontrol degeri eslesiyor.\n")
}

grafik_kaydet <- function(veri) {
  temiz <- hazirla(veri)
  sonuc <- test_hesapla(temiz)
  sinir <- max(5, abs(sonuc["t"]) + 1, sonuc["kritik_t"] + 1)
  eksen <- seq(-sinir, sinir, length.out = 2001)
  yogunluk <- dt(eksen, df = temiz$serbestlik)
  dir.create("ciktilar/r", recursive = TRUE, showWarnings = FALSE)
  png("ciktilar/r/test-ve-aralik.png", width = 1650, height = 600, res = 150)
  on.exit(dev.off())
  par(mfrow = c(1, 2))
  plot(eksen, yogunluk, type = "l", col = "#1F4E79", xlab = "H0 altinda t", ylab = "Yogunluk",
       main = paste("Cift yonlu p =", format(sonuc["p_cift"], digits = 5)))
  for (indis in list(eksen <= -abs(sonuc["t"]), eksen >= abs(sonuc["t"]))) {
    polygon(c(eksen[indis][1], eksen[indis], tail(eksen[indis], 1)),
            c(0, yogunluk[indis], 0), col = "#8FB8D8", border = NA)
  }
  abline(v = c(-abs(sonuc["t"]), abs(sonuc["t"])), lty = 3, col = "#1F4E79")
  aralik <- sonuc[c("alt_sinir", "ust_sinir")]
  sinirlar <- range(c(aralik, temiz$null_degeri)) + c(-.3, .3) * sonuc["hata_payi"]
  plot(NA, xlim = sinirlar, ylim = c(-.5, .7), yaxt = "n", ylab = "",
       xlab = "Evren farki delta (fark birimi)", main = paste("Guven duzeyi:", 1 - temiz$alfa))
  segments(aralik[1], 0, aralik[2], 0, col = "#087F5B", lwd = 2)
  points(temiz$fark, 0, pch = 19, col = "#087F5B")
  abline(v = temiz$null_degeri, lty = 2, col = "#555555")
  for (uc in aralik) {
    segments(uc, -.06, uc, .06, col = "#087F5B")
    text(uc, .2, sprintf("%.4f", uc))
  }
  cat("Grafik: ciktilar/r/test-ve-aralik.png\n")
}

secenek <- commandArgs(trailingOnly = TRUE)
if (any(!secenek %in% c("--check", "--grafik"))) stop("Bilinmeyen secenek.")
veri <- read.csv("veri.csv", check.names = FALSE)
sonuc <- hesapla(veri)
print(sonuc, row.names = FALSE)
cat("Ana cift yonlu karar:", if (test_hesapla(veri)["reddet_cift"] == 1) "H0 reddedilir" else "H0 reddedilemez", "\n")
if ("--check" %in% secenek) kontrol_et(sonuc, read.csv("beklenen-sonuclar.csv", stringsAsFactors = FALSE))
if ("--grafik" %in% secenek) grafik_kaydet(veri)
sessionInfo()

File: companion/bolumler/b09/ornek-01/cozum.py
import argparse
import math
from pathlib import Path
import platform

import numpy as np
import pandas as pd
import scipy
from scipy import stats


def hazirla(veri):
    if list(veri.columns) != ["fark", "standart_hata", "serbestlik", "null_degeri", "alfa"] or len(veri) != 1:
        raise ValueError("Tek satirda fark, standart_hata, serbestlik, null_degeri, alfa ozeti gerekli.")
    if veri.isna().any().any():
        raise ValueError("Eksik ozet kabul edilmez.")
    for sutun in veri.columns:
        if not pd.api.types.is_numeric_dtype(veri[sutun]) or pd.api.types.is_bool_dtype(veri[sutun]):
            raise ValueError("Sayisal ozet degerleri gerekli.")
    if not np.isfinite(veri.to_numpy(dtype=float)).all():
        raise ValueError("Sonlu ozet degerleri gerekli.")
    temiz = veri.reset_index(drop=True).copy(deep=True)
    if temiz.loc[0, "standart_hata"] <= 0 or temiz.loc[0, "serbestlik"] <= 0:
        raise ValueError("Standart hata ve serbestlik pozitif olmali.")
    if not 0 < temiz.loc[0, "alfa"] < 1:
        raise ValueError("Alfa 0 ile 1 arasinda olmali; 5 yerine 0.05 kullanin.")
    return temiz


def test_hesapla(veri):
    temiz = hazirla(veri)
    fark, standart_hata, serbestlik, null_degeri, alfa = map(float, temiz.iloc[0])
    t_degeri = (fark - null_degeri) / standart_hata
    p_cift = float(2 * stats.t.sf(abs(t_degeri), df=serbestlik))
    p_ust = float(stats.t.sf(t_degeri, df=serbestlik))
    p_alt = float(stats.t.cdf(t_degeri, df=serbestlik))
    kritik = float(stats.t.ppf(1 - alfa / 2, df=serbestlik))
    hata_payi = kritik * standart_hata
    alt_sinir, ust_sinir = fark - hata_payi, fark + hata_payi
    sonuc = {"t": t_degeri, "p_cift": p_cift, "p_ust": p_ust, "p_alt": p_alt,
             "reddet_cift": int(p_cift < alfa), "reddet_alfa001": int(p_cift < 0.01),
             "reddet_ust": int(p_ust < alfa), "reddet_alt": int(p_alt < alfa),
             "guven_duzeyi": 1 - alfa, "kritik_t": kritik, "hata_payi": hata_payi,
             "alt_sinir": alt_sinir, "ust_sinir": ust_sinir, "genislik": 2 * hata_payi,
             "null_aralikta": int(alt_sinir <= null_degeri <= ust_sinir)}
    if not all(math.isfinite(value) for value in sonuc.values()):
        raise ValueError("Test veya aralik sayisal olarak hesaplanamadi.")
    return sonuc


def hesapla(veri):
    temiz = hazirla(veri)
    sonuc = test_hesapla(temiz)
    satirlar = [("girdi", sutun, float(temiz.loc[0, sutun])) for sutun in temiz.columns]
    satirlar.extend(("test", anahtar, sonuc[anahtar]) for anahtar in
                    ["t", "p_cift", "p_ust", "p_alt", "reddet_cift", "reddet_alfa001", "reddet_ust", "reddet_alt"])
    satirlar.extend(("aralik", anahtar, sonuc[anahtar]) for anahtar in
                    ["guven_duzeyi", "kritik_t", "hata_payi", "alt_sinir", "ust_sinir", "genislik", "null_aralikta"])
    return pd.DataFrame(satirlar, columns=["degisken", "olcu", "deger"])


def kontrol_et(sonuc, beklenen):
    if list(beklenen.columns) != ["degisken", "olcu", "deger"]:
        raise ValueError("Kontrol sutunlari uyusmuyor.")
    if not sonuc[["degisken", "olcu"]].equals(beklenen[["degisken", "olcu"]]):
        raise ValueError("Kontrol etiketleri veya sirasi uyusmuyor.")
    hedefler = pd.to_numeric(beklenen["deger"], errors="raise")
    if not np.isfinite(hedefler).all() or not np.isfinite(sonuc["deger"]).all():
        raise ValueError("Sonlu kontrol degerleri gerekli.")
    for gercek, hedef in zip(sonuc["deger"], hedefler):
        if not math.isclose(gercek, hedef, rel_tol=1e-9, abs_tol=1e-9):
            raise ValueError(f"Kontrol uyusmazligi: {gercek} != {hedef}")
    print(f"DOGRULANDI: {len(sonuc)} kontrol degeri eslesiyor.")


def grafik_kaydet(veri):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    temiz = hazirla(veri)
    sonuc = test_hesapla(temiz)
    fark = float(temiz.loc[0, "fark"])
    null_degeri = float(temiz.loc[0, "null_degeri"])
    serbestlik = float(temiz.loc[0, "serbestlik"])
    sinir = max(5, abs(sonuc["t"]) + 1, sonuc["kritik_t"] + 1)
    eksen = np.linspace(-sinir, sinir, 2001)
    yogunluk = stats.t.pdf(eksen, df=serbestlik)
    fig, paneller = plt.subplots(1, 2, figsize=(11, 4))
    paneller[0].plot(eksen, yogunluk, color="#1F4E79")
    paneller[0].fill_between(eksen, yogunluk, where=np.abs(eksen) >= abs(sonuc["t"]), color="#8FB8D8")
    for konum in [-abs(sonuc["t"]), abs(sonuc["t"])]:
        paneller[0].axvline(konum, color="#1F4E79", linestyle=":")
    paneller[0].set(xlabel="H0 altinda t", ylabel="Yogunluk", title=f'Cift yonlu p = {sonuc["p_cift"]:.5f}')
    paneller[1].errorbar(fark, 0, xerr=sonuc["hata_payi"], fmt="o", capsize=7, color="#087F5B")
    paneller[1].axvline(null_degeri, color="#555555", linestyle="--", label=f"H0: delta={null_degeri:g}")
    for uc in [sonuc["alt_sinir"], sonuc["ust_sinir"]]:
        paneller[1].annotate(f"{uc:.4f}", (uc, 0), xytext=(0, 12), textcoords="offset points", ha="center")
    paneller[1].set(xlim=(min(null_degeri, sonuc["alt_sinir"]) - sonuc["hata_payi"] * .3,
                         max(null_degeri, sonuc["ust_sinir"]) + sonuc["hata_payi"] * .3),
                    yticks=[], ylim=(-.5, .7), xlabel="Evren farki delta (fark birimi)",
                    title=f'%{100 * sonuc["guven_duzeyi"]:g} iki tarafli guven araligi')
    paneller[1].legend(loc="lower right", fontsize="small")
    fig.tight_layout()
    hedef = Path("ciktilar/python")
    hedef.mkdir(parents=True, exist_ok=True)
    fig.savefig(hedef / "test-ve-aralik.png", dpi=150)
    plt.close(fig)
    print("Grafik: ciktilar/python/test-ve-aralik.png")


def main():
    parser = argparse.ArgumentParser(description="B09: ozetten hipotez testi ve guven araligi")
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--grafik", action="store_true")
    secenek = parser.parse_args()
    veri = pd.read_csv("veri.csv")
    sonuc = hesapla(veri)
    print(sonuc.to_string(index=False))
    karar = test_hesapla(veri)["reddet_cift"]
    print("Ana cift yonlu karar:", "H0 reddedilir" if karar else "H0 reddedilemez")
    if secenek.check:
        kontrol_et(sonuc, pd.read_csv("beklenen-sonuclar.csv"))
    if secenek.grafik:
        grafik_kaydet(veri)
    print(f"Python {platform.python_version()}, NumPy {np.__version__}, pandas {pd.__version__}, SciPy {scipy.__version__}")


if __name__ == "__main__":
    main()